src/main/java/com/exasol/spark/s3/ExasolS3ScanBuilder.java

```diff
@@ -0,0 +1,203 @@
+ */
+public class ExasolS3ScanBuilder implements ScanBuilder, SupportsPushDownFilters, SupportsPushDownRequiredColumns {
+    private static final Logger LOGGER = Logger.getLogger(ExasolS3ScanBuilder.class.getName());
+    private final ExasolOptions options;
+    private final CaseInsensitiveStringMap properties;
+
+    private StructType schema;
+    private Filter[] pushedFilters;
+
+    /**
+     * Creates a new instance of {@link ExasolS3ScanBuilder}.
+     *
+     * @param options    user provided options
+     * @param schema     user-provided {@link StructType} schema
+     * @param properties original key-value properties map that is passed to delegating classes
+     */
+    public ExasolS3ScanBuilder(final ExasolOptions options, final StructType schema,
+            final CaseInsensitiveStringMap properties) {
+        this.options = options;
+        this.schema = schema;
+        this.properties = properties;
+        this.pushedFilters = new Filter[0];
+    }
+
+    @Override
+    public Filter[] pushFilters(final Filter[] filters) {
+        final List<Filter> unsupportedFilters = getUnsupportedFilters(filters);
+        final List<Filter> supportedFilters = new ArrayList<>(Arrays.asList(filters));
+        supportedFilters.removeAll(unsupportedFilters);
+        this.pushedFilters = supportedFilters.toArray(new Filter[] {});
+        return unsupportedFilters.toArray(new Filter[] {});
+    }
+
+    private List<Filter> getUnsupportedFilters(final Filter[] filters) {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Filter[] pushedFilters() {
+        return this.pushedFilters;
+    }
+
+    @Override
+    public void pruneColumns(final StructType schema) {
+        this.schema = schema;
+    }
+
+    @Override
+    public Scan build() {
+        final SparkSession sparkSession = SparkSession.active();
+        final String s3Bucket = this.options.getS3Bucket();
+        final String s3BucketKey = UUID.randomUUID() + "-" + sparkSession.sparkContext().applicationId();
+        // Import query data into `s3Bucket/s3BucketKey` location as `CSV` files
+        importDataIntoS3Location(sparkSession, s3Bucket, s3BucketKey);
+        // Uses Spark `CSVTable` to read `CSV` files
+        final Seq<String> csvFilesPaths = getS3CSVFiles(s3Bucket, s3BucketKey);
+        return new CSVTable("", sparkSession, this.properties, csvFilesPaths, Option.apply(this.schema),
+                new CSVFileFormat().getClass()).newScanBuilder(getUpdatedMapWithCSVOptions(this.properties)).build();
+    }
+
+    private void importDataIntoS3Location(final SparkSession spark, final String s3Bucket, final String s3BucketKey) {
+        LOGGER.info(() -> "Using S3 bucket '" + s3Bucket + "' with folder '" + s3BucketKey + "' for scan job data.");
+        prepareIntermediateData(s3BucketKey);
+    }
+
+    private Seq<String> getS3CSVFiles(final String s3Bucket, final String s3BucketKey) {
```

Comment:
Proposal: rename variable and method to `getCsvFiles()`?
Arguments bucket and key already indicate that the files are retrieved from S3 in my opinion. And also the whole package is named `s3` anyway.

Reply:
Removed also in other parts